cabal-testsuite/src/Test/Cabal/Prelude.hs

```diff
@@ -937,6 +949,55 @@ isJavaScript = buildArch == JavaScript
 skipIfWindows :: String -> IO ()
 skipIfWindows why = skipIfIO ("Windows " <> why) isWindows
 
+skipIfOSX :: String -> IO ()
+skipIfOSX why = skipIfIO ("OSX " <> why) isOSX
+
+skipIfCI :: Int -> IO ()
+skipIfCI ticket = skipIfIO ("CI, see #" <> show ticket) =<< isCI
+
+skipIfCIAndWindows :: Int -> IO ()
+skipIfCIAndWindows ticket = skipIfIO ("Windows CI, see #" <> show ticket) . (isWindows &&) =<< isCI
+
+skipIfCIAndOSX :: Int -> IO ()
+skipIfCIAndOSX ticket = skipIfIO ("OSX CI, see #" <> show ticket) . (isOSX &&) =<< isCI
+
+expectBrokenIfWindows :: Int -> TestM a -> TestM a
+expectBrokenIfWindows ticket = expectBrokenIf isWindows ticket
+
+expectBrokenIfWindowsCI :: Int -> TestM a -> TestM a
+expectBrokenIfWindowsCI ticket m = do
+    ci <- liftIO isCI
+    expectBrokenIf (isWindows && ci) ticket m
+
+expectBrokenIfWindowsCIAndGhc :: String -> Int -> TestM a -> TestM a
+expectBrokenIfWindowsCIAndGhc range ticket m = do
+    ghcVer <- isGhcVersion range
+    ci <- liftIO isCI
+    expectBrokenIf (isWindows && ghcVer && ci) ticket m
+
+expectBrokenIfWindowsAndGhc :: String -> Int -> TestM a -> TestM a
+expectBrokenIfWindowsAndGhc range ticket m = do
+    ghcVer <- isGhcVersion range
+    expectBrokenIf (isWindows && ghcVer) ticket m
+
+expectBrokenIfOSXAndGhc :: String -> Int -> TestM a -> TestM a
+expectBrokenIfOSXAndGhc range ticket m = do
+    ghcVer <- isGhcVersion range
+    expectBrokenIf (isOSX && ghcVer) ticket m
+
+expectBrokenIfGhc :: String -> Int -> TestM a -> TestM a
+expectBrokenIfGhc range ticket m = do
+    ghcVer <- isGhcVersion range
+    expectBrokenIf ghcVer ticket m
+
+flakyIfCI :: Int -> TestM a -> TestM a
+flakyIfCI ticket m = do
+    ci <- liftIO isCI
+    flakyIf ci ticket m
+
+flakyIfWindows :: Int -> TestM a -> TestM a
+flakyIfWindows ticket m = flakyIf isWindows ticket m
```

Comment:
I am not sure I buy it as an improvement having separate functions for so many combinations. It'd be great to have a little DSL for expressing such things and this looks like a step away from that. But it is tempting to go this route because it does save on typing. Perhaps if Haskell had a built-in the monadic bang (in the sense of [this plugin](https://hackage.haskell.org/package/monadic-bang)), the savings here wouldn't be as noticeable. So, overall, I don't feel strongly about it, just curious what others think.

Reply:
Yeah, as I was writing them I was thinking the same. But overall I think it is useful to make the skip and broken messages as uniform as possible, to be able to grep the output. 

Otherwise one might use "shared libs", another person uses "dynamic libs", another one makes a typo "shred libs", etc.

I will consider whether to add something like a DSL. But maybe this would be mergeable? At the very least I could put a comment in the code saying "TODO: this should be reworked into a DSL or some such".

Reply:
You don't need to risk loss of uniformity. See for example https://hackage.haskell.org/package/xmonad-contrib-0.18.0/docs/XMonad-Util-WindowProperties.html.
In this case I'd have 4 parameters: expected state or action (skip, expect fail, expect pass, flaky…), predicate (cf. above WindowProperties, replacing your multiple conditionals), ticket, message. But this might indeed be future work.

Reply:
Ah i didn't explain myself. I think a DSL would be uniform and the best solution. What is not uniform is
```
skipIf noShared "no shared libs"
```
And somewhere else
```
skipIf noShared "no dyn libs"
```

Which is what was there before I introduced these functions.

Reply:
I would make that aspect of the message come from the DSL, and the explanation should give details. (In which case perhaps it should be optional, or absorbed into the ticket.)

Reply:
Another option I could see is by making the combinators return a `Maybe String` and then `skipIf` would unpack it if `Just`, something like

```
isWindows :: Maybe String
isWindows = if buildOS == Windows then Just "Windows" else Nothing

skipIf mSkipMsg test = maybe test skip mSkipMsg
```

but that is counterintuitive. The suffix `if` usually is interpreted as expecting a `Bool`...

Reply:
Ok I managed to make the CI green. From my side there are no other blockers for merging this.

What is your opinion on the combinators then @ulysses4ever @geekosaur ? Could this be merged and the DSL be done in a later PR or do you think this should not be merged before the DSL is done?

Reply:
By no means I meant that the DSL should be a prerequisite.

I simply struggle to get to the bottom of the long line of commits. For me personally, it'd be much easier to approve commits in separate PRs one by one. But I understand how inconvenient that approach is for the author. So I'll try to get a review done.

Reply:
Some points I wanted to note down:
- It would have been perfectly ok if you said you wanted the DSL before approving, it is a perfectly valid opinion and I asked because I would have thought it was perfectly ok if you said "I will not give my approval until this PR has better ergonomics for the combinators like a DSL maybe"
- I have tried to split the changes semantically, one on each commit, so that it would be easier to review. Some of them are interesting, some are boring 😄
- (Btw, while reviewing you can choose to see only one commit at a time, that might make it easier)
- I'm fine with splitting the work in multiple PRs if it is independent (kind of like in this case, it could have been 4-5 PRs) but I think it is more work to drag attention to 5 PRs for review than to one. There is usually not enough volunteer-time to review this so I prefer just making it in one go 🙂 (as long as the commits are related, like in this case they are all about CI and tests)
- This being said, if you prefer I can split this PR in multiple PRs.

Reply:
I think I already suggested it should be future work? It wants some up-front design.

Reply:
For splitting: I think you should do it the way you prefer. For me personally reviewing and approving 5 small PRs can take much less time than reviewing 5 commits and approving one PR  with those. That's simply because it's not possible to "save state" (meaning my mental state) in between the commits while there's a way to do that for multiple PRs (by approving them one by one). Unless I'm missing something in GitHub UI. I know that you can review commits separately (that's what I've been doing with this PR), but that's not the same as reviewing/approving/and forgetting about individual PRs (which works better for me). Again, you should do it the way you find it more convenient, I think.

For the DSL, I share Brandon's opinion (see the comment just above) entirely.